File: jsx/src/detectors/sql_injection/sql_injection.js
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

// {fact rule=sql-injection@v1.0 defects=1}
var sql = require('mysql')
var express = require('express')
var app = express()

var connection = sql.createConnection({
    host     : 'localhost',
    user     : 'myUserName',
    password : 'myPass',
    database : 'myDatabase'
})

function sqlInjectionNoncompliant() {
    app.get("/user/:id", (req, res) => {
        // Noncompliant: user input is not sanitized before use.
        var query = "SELECT * FROM Employees WHERE ID = " + req.params.id
        connection.query(query, (error, results, fields) => {
            if (error) throw error
        })
    })
}
// {/fact}


// {fact rule=sql-injection@v1.0 defects=0}
var sql = require('mysql')
var express = require('express')
var app = express()

var connection = sql.createConnection({
    host     : 'localhost',
    user     : 'myUserName',
    password : 'myPass',
    database : 'myDatabase'
})

function sqlInjectionCompliant() {
    app.get("/user/:id", (req, res) => {
        // Compliant: user input is sanitized before use.
        var query = "SELECT * FROM Employees WHERE ID = " + connection.escape(req.params.id)
        connection.query(query, (error, results, fields) => {
            if (error) throw error
        })
    })
}
// {/fact}

File: python/src/detectors/sql_injection/sql_injection.py
#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
#  SPDX-License-Identifier: Apache-2.0

# {fact rule=sql-injection@v1.0 defects=1}
def execute_query_noncompliant(request):
    import sqlite3

    name = request.GET.get("name")
    query = "SELECT * FROM Users WHERE name = " + name + ";"
    with sqlite3.connect("example.db") as connection:
        cursor = connection.cursor()
        # Noncompliant: user input is used without sanitization.
        cursor.execute(query)
        connection.commit()
        connection.close()
# {/fact}


# {fact rule=sql-injection@v1.0 defects=0}
def execute_query_compliant(request):
    import re
    import sqlite3

    name = request.GET.get("name")
    query = "SELECT * FROM Users WHERE name = ?;"
    with sqlite3.connect("example.db") as connection:
        cursor = connection.cursor()
        # Compliant: A parameterized query is used to pass an external input.
        cursor.execute(query, (name,))
        connection.commit()
# {/fact}
